portal/src/main/webapp/js/src/oncoprint/webgl/setup.js

```diff
@@ -1734,6 +1734,46 @@ window.CreateCBioPortalOncoprintWithToolbar = function (ctr_selector, toolbar_se
 		    }
 		    oncoprint.removeTracks(tracks_to_remove);
 		});
+		
+		$(toolbar_selector + ' #oncoprint_diagram_heatmap_menu #cluster_heatmaps_btn').click(function (evt) {
+		    // suppress dropdown hiding and form submit
+		    evt.preventDefault();
+		    evt.stopPropagation();
+
+		    // cluster heatmap for selected profile:
+		    var genetic_profile_id = $(toolbar_selector + ' #oncoprint_diagram_heatmap_menu #oncoprint_diagram_heatmap_profiles').val();
+		    if (genetic_profile_id) {
+				//sort according to order found in the clustering results: 
+				State.sorting_by_given_order = true;
+		    	var grp = State.heatmap_track_groups[genetic_profile_id];
+				var heatmap_track_group_id = grp.track_group_id; 
+				var heatmap_data_deferred = State.using_sample_data ? getSampleHeatmapData(grp.genetic_profile_id, Object.keys(grp.gene_to_track_id)) : QuerySession.getPatientHeatmapData(grp.genetic_profile_id, Object.keys(grp.gene_to_track_id));
+				var case_ids_deferred =  State.using_sample_data ? QuerySession.getSampleIds() : QuerySession.getPatientIds();
+				//process data, call clustering:
+				$.when(QuerySession.getCaseUIDMap(), QuerySession.getTrackUIDMap(heatmap_track_group_id, "GENE_HEATMAP_DATA"), heatmap_data_deferred, case_ids_deferred).then(
+						function (case_uid_map, track_uid_map, heatmap_data, case_ids) {
+							$.when(QuerySession.getClusteringOrder(case_uid_map, track_uid_map, heatmap_data, case_ids)).then(
+									function (clusteringResult) {
+										LoadingBar.update(0.9);
+										oncoprint.setSortConfig({'type': 'order', order: clusteringResult.sampleUidsInClusteringOrder});
+										oncoprint.setTrackGroupOrder(heatmap_track_group_id, clusteringResult.entityUidsInClusteringOrder);
+										LoadingBar.hide();
+									});
+							//show progress bar:
+							LoadingBar.show();
+							LoadingBar.update(0.2);
+						    LoadingBar.msg("Clustering...");
+						    //update progress bar every 0.2s:
+						    window.setInterval(function() {
+						    	var d = new Date();
+						    	var n = d.getMilliseconds();
+							    LoadingBar.update(n/1000);
```

Comment:
This is a nitpick but I'm not crazy about this because normally the loading bar size actually signifies progress percentage. If you're going to do this, I'd rather show a different color than green.